File: leetcode/python/p1769.py
from typing import List


class Solution:
    """
    https://leetcode.com/problems/minimum-number-of-operations-to-move-all-balls-to-each-box
    """

    def minOperationsn2(self, boxes: str) -> List[int]:
        n = len(boxes)
        res = [0] * n
        for i in range(n):
            for j in range(n):
                if boxes[j] == '1':
                    res[i] += abs(i - j)
        return res

    def minOperations(self, boxes: str) -> List[int]:
        n = len(boxes)
        answer = [0] * n
        balls_to_left = 0
        moves_to_left = 0
        balls_to_right = 0
        moves_to_right = 0
        for i in range(len(boxes)):
            answer[i] += moves_to_left
            balls_to_left += int(boxes[i])
            moves_to_left += balls_to_left

        for i in range(len(boxes) - 1, -1, -1):
            answer[i] += moves_to_right
            balls_to_right += int(boxes[i])
            moves_to_right += balls_to_right

        return answer


if __name__ == "__main__":
    s = Solution()
    print(s.minOperations("110"))

File: leetcode/p1769.cpp


class Solution {
   public:
    std::vector<int> minOperationsn2(std::string boxes) {
        int n = boxes.length();
        std::vector<int> res(n, 0);
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (boxes[j] == '1') {
                    res[i] += abs(i - j);
                }
            }
        }
        return res;
    }

    std::vector<int> minOperations(std::string boxes) {
        int n = boxes.length();
        std::vector<int> answer(n, 0);
        int balls_to_left = 0;
        int moves_to_left = 0;
        int balls_to_right = 0;
        int moves_to_right = 0;

        for (int i = 0; i < n; i++) {
            answer[i] += moves_to_left;
            balls_to_left += (boxes[i] - '0');
            moves_to_left += balls_to_left;
        }

        for (int i = n - 1; i >= 0; i--) {
            answer[i] += moves_to_right;
            balls_to_right += (boxes[i] - '0');
            moves_to_right += balls_to_right;
        }

        return answer;
    }
};
